File: packages/node/iii-example/src/streams.ts
import { iii } from './iii'
import type { Todo } from './types'

export const streams = {
  get: async (stream_name: string, group_id: string, item_id: string): Promise<any | null> => {
    return iii.call('streams.get', { stream_name, group_id, item_id })
  },
  set: async (stream_name: string, group_id: string, item_id: string, data: any): Promise<any> => {
    return iii.call('streams.set', { stream_name, group_id, item_id, data })
  },
  delete: async (stream_name: string, group_id: string, item_id: string): Promise<void> => {
    return iii.call('streams.delete', { stream_name, group_id, item_id })
  },
  getGroup: async (stream_name: string, group_id: string): Promise<any[]> => {
    return iii.call('streams.getGroup', { stream_name, group_id })
  },
  listGroups: async (stream_name: string): Promise<string[]> => {
    return iii.call('streams.listGroups', { stream_name })
  },
}

let todoState: Todo[] = []

iii.createStream('todo', {
  get: async input => todoState.find(todo => todo.id === input.item_id),
  set: async input => {
    const existingTodo = todoState.find(todo => todo.id === input.item_id)

    if (existingTodo) {
      const newTodo = { ...existingTodo, ...input.data }
      todoState = todoState.map(todo => (todo.id === input.item_id ? newTodo : todo))
      return { old_value: existingTodo, new_value: newTodo }
    }

    const newTodo = {
      id: input.item_id,
      groupId: input.group_id,
      description: input.data.description,
      dueDate: input.data.dueDate,
      completedAt: null,
    }

    todoState.push(newTodo)

    return { old_value: undefined, new_value: newTodo }
  },
  delete: async input => {
    const old_value = todoState.find(todo => todo.id === input.item_id)
    todoState = todoState.filter(todo => todo.id !== input.item_id)
    return { old_value }
  },
  getGroup: async input => todoState.filter(todo => todo.groupId === input.group_id),
  listGroups: async () => [...new Set(todoState.map(todo => todo.groupId))],
  update: async () => {
    throw new Error('Not implemented')
  },
})

File: packages/python/iii/src/iii/streams.py
"""Stream types and interfaces for the III SDK."""

from abc import ABC, abstractmethod
from typing import Any, Generic, TypeVar

from pydantic import BaseModel

TData = TypeVar("TData")


class StreamAuthInput(BaseModel):
    """Input for stream authentication."""

    headers: dict[str, str]
    path: str
    query_params: dict[str, list[str]]
    addr: str


class StreamAuthResult(BaseModel):
    """Result of stream authentication."""

    context: Any | None = None


class StreamJoinLeaveEvent(BaseModel):
    """Event for stream join/leave."""

    subscription_id: str
    stream_name: str
    group_id: str
    id: str | None = None
    context: Any | None = None


class StreamJoinResult(BaseModel):
    """Result of stream join."""

    unauthorized: bool


class StreamGetInput(BaseModel):
    """Input for stream get operation."""

    stream_name: str
    group_id: str
    item_id: str


class StreamSetInput(BaseModel):
    """Input for stream set operation."""

    stream_name: str
    group_id: str
    item_id: str
    data: Any


class StreamDeleteInput(BaseModel):
    """Input for stream delete operation."""

    stream_name: str
    group_id: str
    item_id: str


class StreamGetGroupInput(BaseModel):
    """Input for stream get group operation."""

    stream_name: str
    group_id: str


class StreamListGroupsInput(BaseModel):
    """Input for stream list groups operation."""

    stream_name: str


class StreamUpdateInput(BaseModel):
    """Input for stream update operation."""

    stream_name: str
    group_id: str
    item_id: str
    ops: list["UpdateOp"]


class StreamSetResult(BaseModel, Generic[TData]):
    """Result of stream set operation."""

    old_value: TData | None = None
    new_value: TData | None = None


class UpdateSet(BaseModel):
    """Set operation for stream update."""

    type: str = "set"
    path: str
    value: Any


class UpdateIncrement(BaseModel):
    """Increment operation for stream update."""

    type: str = "increment"
    path: str
    by: int | float


class UpdateDecrement(BaseModel):
    """Decrement operation for stream update."""

    type: str = "decrement"
    path: str
    by: int | float


class UpdateRemove(BaseModel):
    """Remove operation for stream update."""

    type: str = "remove"
    path: str


class UpdateMerge(BaseModel):
    """Merge operation for stream update."""

    type: str = "merge"
    path: str
    value: Any


UpdateOp = UpdateSet | UpdateIncrement | UpdateDecrement | UpdateRemove | UpdateMerge


class IStream(ABC, Generic[TData]):
    """Abstract interface for stream operations."""

    @abstractmethod
    async def get(self, input: StreamGetInput) -> TData | None:
        """Get an item from the stream."""
        ...

    @abstractmethod
    async def set(self, input: StreamSetInput) -> StreamSetResult[TData] | None:
        """Set an item in the stream."""
        ...

    @abstractmethod
    async def delete(self, input: StreamDeleteInput) -> None:
        """Delete an item from the stream."""
        ...

    @abstractmethod
    async def get_group(self, input: StreamGetGroupInput) -> list[TData]:
        """Get all items in a group."""
        ...

    @abstractmethod
    async def list_groups(self, input: StreamListGroupsInput) -> list[str]:
        """List all groups in the stream."""
        ...

    @abstractmethod
    async def update(self, input: StreamUpdateInput) -> StreamSetResult[TData] | None:
        """Update an item in the stream."""
        ...
